### PR TITLE
add better workspace dependency handling

### PR DESCRIPTION
**Pull request by planthaber**
As discussed here:
https://github.com/dfki-ric/docker_image_development/issues/9

### CODE REVIEW

File: image_setup/02_devel_image/Dockerfile
Comment (0Nel) [inline image_setup/02_devel_image/Dockerfile:25]: Doesn't this conflict with mounting this file through exec.bash? I get that this is required for the next line, but shouldn't we work on the mounted file version and cat the file on the host system instead? Or do we need the file inside the image to cat it?

Would this work?
```Dockerfile
RUN cat workspace_os_dependencies.txt | xargs sudo apt install -y
```

File: doc/042_Devel_Image.md
Comment (0Nel) [inline doc/042_Devel_Image.md:75]: `Install instructions for packages that are not workspace dependencies (or in the metadata) sould be directly added into the dockerfile`

Directly into the dockerfile or by using `workspace_os_dependencies.txt`?

I think to avoid confusion, both ways should work. Also adding write_osdeps.bash as a startscript would be handy imo :+1:

File: doc/042_Devel_Image.md
Comment (0Nel) [inline doc/042_Devel_Image.md:72]: Minor typo:

~scripts help~ => script

File: doc/042_Devel_Image.md
Comment (0Nel) [inline doc/042_Devel_Image.md:73]: Minor typo:

~is~ available. => , if available.

**Review by 0Nel — COMMENTED** [review]
I like this approach a lot! I think this makes the whole os-dependency management more clear and convenient. Main point for me atm is mounting and adding the workspace_os_dependency.txt file. Does that work? Is that necessary? More details in the comments ;)

File: image_setup/02_devel_image/Dockerfile
Comment (planthaber) [reply image_setup/02_devel_image/Dockerfile:12]: right, i guess it is a merge leftover...

File: image_setup/02_devel_image/write_osdeps.bash
Comment (planthaber) [reply image_setup/02_devel_image/write_osdeps.bash:7]: There should no manual packages in that file, I thought manual files should be added into the Dockerfile directly.
Perhaps that could be mentioned as comment at the top of the file

File: image_setup/02_devel_image/write_osdeps.bash
Comment (planthaber) [reply image_setup/02_devel_image/write_osdeps.bash:5]: Didn't test ROS yet, i think we should just remove the cd command completely and let the user switch to the correct folder manually

Comment (planthaber) [reply image_setup/02_devel_image/write_osdeps.bash:5]: and if no scritp is called print a note where the script should be called

File: image_setup/02_devel_image/Dockerfile
Comment (planthaber) [reply image_setup/02_devel_image/Dockerfile:25]: There is no conflict in that, when the file is mounted the mounted version is used, RUN is executed in a container, so workspace_os_dependencies.txt must be available. If we want to be sure, we could:

    RUN cat workspace_os_dependencies.txt | xargs sudo apt install -y && rm -f workspace_os_dependencies.txt

File: doc/042_Devel_Image.md
Comment (planthaber) [reply doc/042_Devel_Image.md:75]: we could just add /opt/ to the PATH? then also setup workspace, etc would work that way

File: image_setup/02_devel_image/Dockerfile
Comment (planthaber) [reply image_setup/02_devel_image/Dockerfile:29]: is there a .sh missing? I thought we already did that

File: image_setup/02_devel_image/write_osdeps.bash
Comment (0Nel) [reply image_setup/02_devel_image/write_osdeps.bash:5]: Why not use the find logic? cd into /opt/workspace is save and from there we'd have a solution. I think it is preferable to write it in a way that doesn't require user interaction. Especially if we would add this as a startscript, which I am really voting for ;)

File: image_setup/02_devel_image/write_osdeps.bash
Comment (planthaber) [reply image_setup/02_devel_image/write_osdeps.bash]: I added a disclaimer in that generated file

File: image_setup/02_devel_image/write_osdeps.bash
Comment (planthaber) [reply image_setup/02_devel_image/write_osdeps.bash]: updated to use find logic, is even better in case someone has more than one workspace installed

File: image_setup/02_devel_image/Dockerfile
Comment (planthaber) [reply image_setup/02_devel_image/Dockerfile]: renamed the ADDed file to installed_workspace_os_dependencies.txt, which 
 now allows to see what is actually installed and to compare with a newly generated file to see if a image update is needed

File: doc/042_Devel_Image.md
Comment (planthaber) [reply doc/042_Devel_Image.md:75]: added to PATH using the Dockerfile

Comment (planthaber) [reply doc/042_Devel_Image.md:75]: and removed it again, you can call it anyway by using `./exec /opt/write_osedps.bash`

**Comment by planthaber** [comment]
In the current state, when building a release we could just call `./exec devel /opt/write_osdeps.bash` from the build script and use the newly generated file in the release Dockerfile similar to the devel dockerfile: https://github.com/dfki-ric/docker_image_development/blob/78ce1ea7d5dc5b3455952f1794ba60e99ec7cb35/image_setup/02_devel_image/Dockerfile#L25

File: image_setup/02_devel_image/workspace_os_dependencies.txt
Comment (0Nel) [inline image_setup/02_devel_image/workspace_os_dependencies.txt]: renamed tools to packages. The tools in combination with editors did confuse me a little..

File: image_setup/02_devel_image/Dockerfile
Comment (0Nel) [reply image_setup/02_devel_image/Dockerfile]: Just in the comments. I edited this with my last commit.

**Review by 0Nel — APPROVED** [review]
Added check if src or autoproj exists in workspace before executing write_osdeps.bash

Fixed DOS style carriage returns that broke filecompare with cpm.

On another note, are mounted files not editable anymore on the host? If I update the file in /image_setup/02_devel_image on my host the edit is only visible inside the container if a new container is created. Is that standard behaviour when mounting files directly? On the one hand that makes sense since new directory mounts are also only re-evaluated upon rebuild. However this makes working with mounted files kind of unintuitive. Still better than copying it, but seemed kind of odd to me.

**Comment by 0Nel** [comment]
Already resolved merge conflicts in another tab. I'll merge as soon as I have your approval.

@planthaber

**Comment by planthaber** [comment]
I just tested editing the mounted workspace_os_dependencies.txt fiel and for me it works both ways, only I needed to force my editor on the host toe re-load the file

**Pull request merged**
